### PR TITLE
Fix typing for most of the AST module

### PR DESCRIPTION
**Pull request by dcbaker**
What remains is more complicated because of layering violations that will be difficult to fix

### CODE REVIEW

File: mesonbuild/ast/printer.py [hunk 1]
```diff
@@ -5,6 +5,8 @@
 # or an interpreter-based tool
 from __future__ import annotations
 
+from attr import has
```
Comment (eli-schwartz) [inline mesonbuild/ast/printer.py:8]: Where does this come from??

Comment (dcbaker) [reply mesonbuild/ast/printer.py:8]: ... I deleted that so many times. My editor seems to think I need that for some reason I can't fathom

**Review by tristan957 — APPROVED** [review]

File: mesonbuild/ast/printer.py [hunk 2]
```diff
@@ -65,8 +65,7 @@ def visit_NumberNode(self, node: mparser.NumberNode) -> None:
         node.lineno = self.curr_line or node.lineno
 
     def escape(self, val: str) -> str:
-        return val.translate(str.maketrans({'\'': '\\\'',
-                                            '\\': '\\\\'}))
+        return val.translate(str.maketrans(T.cast(T.Dict[str, T.Union[str, int]], {'\'': '\\\'', '\\': '\\\\'})))
```
Comment (eli-schwartz) [inline mesonbuild/ast/printer.py:68]: First argument to T.cast should be quoted.

File: run_mypy.py [hunk 3]
```diff
@@ -26,6 +26,7 @@
     # specific files
     'mesonbuild/ast/printer.py',
     'mesonbuild/ast/postprocess.py',
+    'mesonbuild/ast/visitor.py',
```
Comment (eli-schwartz) [inline run_mypy.py:30]: I'd probably merge marking multiple files as mypy-clean, into one commit, personally.

**Review by tristan957 — APPROVED** [review]

File: mesonbuild/ast/introspection.py [hunk 4]
```diff
@@ -270,7 +271,7 @@ def traverse_nodes(inqueue: T.List[BaseNode]) -> T.List[BaseNode]:
         empty_sources: T.List[T.Any] = []
         # Passing the unresolved sources list causes errors
         kwargs_reduced['_allow_no_sources'] = True
-        target = targetclass(name, self.subdir, self.subproject, for_machine, empty_sources, [], objects,
+        target = targetclass(name, self.subdir, self.subproject, for_machine, empty_sources, None, objects,
```
Comment (eli-schwartz) [inline mesonbuild/ast/introspection.py:274]: Oh fun... `if self.structured_sources` and an empty list is falsey.

**Pull request merged**